backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/projection/impl/DistanceFieldSearchProjectionImpl.java

```diff
@@ -40,18 +46,44 @@
 	}
 
 	@Override
-	public void contributeRequest(JsonObject requestBody) {
-		SCRIPT_FIELDS_ACCESSOR
-				.property( scriptFieldName ).asObject()
-				.property( "script" ).asObject()
-				.set( requestBody, createScript( absoluteFieldPath, center ) );
+	public void contributeRequest(JsonObject requestBody, ElasticsearchSearchQueryElementCollector elementCollector) {
+		this.distanceSortIndex = elementCollector.getDistanceSortIndex( absoluteFieldPath, center );
```

Comment:
Introducing mutability here is not a great idea, even if we consider `SearchProjections` should not be shared between multiple threads.

Try this:

```
SearchTarget target =...;
GeoPoint center = ...;
SearchProjection<Double> distanceProj = target.projection().distance( "field1", center ).toProjection();

SearchQuery query1 = target.query().asProjections( distanceProj ).sort().distance( "field1", center ).then().distance( "field2", center ).build();
SearchQuery query2 = target.query().asProjections( distanceProj ).sort().distance( "field2", center ).then().distance( "field1", center ).build();

SearchResult<?> resultsOfQuery1 = query1.execute();
SearchResult<?> resultsOfQuery2 = query2.execute();
```

I bet the results of `query1` will be wrong, because the sort indexes will have been overridden by the creation of `query2`.

You need to avoid mutating the `SearchProjection`, and either make the `ElementCollector` accessible to the `extract()` method, or create another object from the element collector that will be accessible from the `extract()` method (`SearchProjectionExecutionContext`, maybe?)

Alternatively, you can clone the projection for each query, or make the `contributeRequest` method return an object to be used for extraction (in most hit extractors, we'll return `this`, but in this case we'll return a different object).

Reply:
Good catch. Introduced a `SearchProjectionExecutionContext` as you suggested.